### PR TITLE
Add tag add and delete command

### LINKED ISSUE #29: Add tags (bulk) to multiple friends
Having existing tags, I would like to add multiple friends to the existing tag quickly.

### PR DESCRIPTION
**Pull request by tanboonji**
Closes #29 

@oeiyiping Docs & Commands
@yaowei-soc Docs & Tests

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DeleteTagCommand.java [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        return new Person(updatedName, updatedPhone, updatedEmail, updatedCompany, updatedJobTitle,
+                updatedAddress, updatedRemark, updatedTags);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof DeleteTagCommand // instanceof handles nulls
+                && tags.equals(((DeleteTagCommand) other).tags)
```
Comment (yaowei-soc) [inline src/main/java/seedu/address/logic/commands/DeleteTagCommand.java:183]: Better to use `containsAll`. See https://github.com/AY2021S2-CS2103T-T12-3/tp/pull/129/commits/bb6c1b9119e98e5f757b1811b8712731482ac534 with corresponding tests https://github.com/AY2021S2-CS2103T-T12-3/tp/pull/129/commits/3c0d56828bffbc1b266fe65ee53a1f3585137632

File: src/main/java/seedu/address/logic/commands/TagCommand.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+public abstract class TagCommand extends Command {
+
+    public static final String COMMAND_WORD = "tag";
+    public static final String ADD_SUB_COMMAND_WORD = "add";
+    public static final String DELETE_SUB_COMMAND_WORD = "delete";
+    public static final String MESSAGE_NO_SELECTED_PERSON = "You have not selected any person.";
+    public static final String MESSAGE_NO_SHOWN_PERSON = "There are no shown person.";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Add or delete tags of person(s) in the address book. "
```
Comment (yaowei-soc) [inline src/main/java/seedu/address/logic/commands/TagCommand.java:14]: Missing `\n` behind

File: src/main/java/seedu/address/logic/commands/TagCommand.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            + "Parameters: "
+            + "SUB_COMMAND [PARAMETERS]\n"
+            + "Sub Commands: "
+            + ADD_SUB_COMMAND_WORD + " INDEX ... [-t] TAG ..., "
```
Comment (yaowei-soc) [inline src/main/java/seedu/address/logic/commands/TagCommand.java:18]: Unicode ellipsis?

File: src/test/java/seedu/address/logic/commands/AddTagCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    @Test
+    public void equals() {
+        AddTagCommand addTagToTargetIndexes = AddTagCommand.createWithTargetIndexes(
+                new ArrayList<>(), getTypicalTags());
+        AddTagCommand addTagToShownIndex = AddTagCommand.createWithShownIndex(getTypicalTags());
+        AddTagCommand addTagToSelectedIndex = AddTagCommand.createWithSelectedIndex(getTypicalTags());
+
+        // same object -> returns true
+        assertEquals(addTagToTargetIndexes, addTagToTargetIndexes);
```
Comment (yaowei-soc) [inline src/test/java/seedu/address/logic/commands/AddTagCommandTest.java:164]: Might want to check different Set<> equals here. Example:

```java
        // different indexes -> not equals
        assertNotEquals(DeleteCommand
                        .buildDeleteIndexCommand(Collections.singletonList(INDEX_FIRST_PERSON)),
                DeleteCommand.buildDeleteIndexCommand(VALID_INDEXES));
```

File: docs/UserGuide.md [hunk 5]
```diff
@@ -108,7 +108,7 @@ Displays the entire User Guide for ease of reference.
 
 Adds a person to the address book.
 
-**Format**: `add -n NAME -p PHONE_NUMBER -e EMAIL -c COMPANY -j JOB_TITLE -a ADDRESS [-r REMARK] [-t TAG]…`
+**Format**: `add -n NAME -p PHONE_NUMBER -e EMAIL -a ADDRESS [-r REMARK] [-t TAG]…`
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply docs/UserGuide.md]: Added it back.

File: src/main/java/seedu/address/logic/commands/AddTagCommand.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+import seedu.address.model.person.Phone;
+import seedu.address.model.person.Remark;
+import seedu.address.model.tag.Tag;
+
+/**
+ * Add tags to person in the address book.
+ */
+public class AddTagCommand extends TagCommand {
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Add tags to person(s) in the address book. "
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply src/main/java/seedu/address/logic/commands/AddTagCommand.java:32]: Updated code for this issue.

File: src/main/java/seedu/address/logic/commands/AddTagCommand.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            + "Parameters: " + ADD_SUB_COMMAND_WORD + " INDEX ... [-t] TAG ...\n"
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply src/main/java/seedu/address/logic/commands/AddTagCommand.java:33]: Updated code for this issue.

File: src/main/java/seedu/address/logic/commands/AddTagCommand.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        return new Person(updatedName, updatedPhone, updatedEmail, updatedCompany, updatedJobTitle,
+                updatedAddress, updatedRemark, updatedTags);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof AddTagCommand // instanceof handles nulls
+                && tags.equals(((AddTagCommand) other).tags)
+                && targetIndexes.equals(((AddTagCommand) other).targetIndexes)
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply src/main/java/seedu/address/logic/commands/AddTagCommand.java:183]: Updated code with this suggestion.

File: src/main/java/seedu/address/logic/commands/DeleteTagCommand.java (hunk 1 above)
Comment (tanboonji) [reply src/main/java/seedu/address/logic/commands/DeleteTagCommand.java:183]: Updated code with this suggestion.

File: src/main/java/seedu/address/logic/commands/TagCommand.java (hunk 2 above)
Comment (tanboonji) [reply src/main/java/seedu/address/logic/commands/TagCommand.java:14]: Updated code for this issue.

File: src/main/java/seedu/address/logic/commands/TagCommand.java (hunk 3 above)
Comment (tanboonji) [reply src/main/java/seedu/address/logic/commands/TagCommand.java:18]: Updated code for this issue.

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 9]
```diff
@@ -88,6 +89,9 @@ public Command parseCommand(String userInput, ReadOnlyUniqueAliasMap aliases) th
         case AliasCommand.COMMAND_WORD:
             return new AliasCommandParser().parse(arguments);
 
+        case TagCommand.COMMAND_WORD:
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply src/main/java/seedu/address/logic/parser/AddressBookParser.java:92]: Added this test into code.

File: src/main/java/seedu/address/logic/parser/TagCommandParser.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public static String tagsToString(Set<Tag> tags) {
+        StringBuilder builder = new StringBuilder();
+
+        for (Tag tag : tags) {
+            builder.append(tag.tagName + ", ");
+        }
+
+        String returnValue = builder.toString();
+        return returnValue.substring(0, returnValue.length() - 2);
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply src/main/java/seedu/address/logic/parser/TagCommandParser.java:93]: Changed code to use streams.

File: src/test/java/seedu/address/logic/commands/AddTagCommandTest.java (hunk 4 above)
Comment (tanboonji) [reply src/test/java/seedu/address/logic/commands/AddTagCommandTest.java:164]: Added test cases for both different indexes and different tags.

File: src/test/java/seedu/address/logic/commands/DeleteTagCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+    @Test
+    public void equals() {
+        DeleteTagCommand deleteTagFromTargetIndexes = DeleteTagCommand.createWithTargetIndexes(
+                new ArrayList<>(), getTypicalTags());
+        DeleteTagCommand deleteTagFromShownIndex = DeleteTagCommand.createWithShownIndex(getTypicalTags());
+        DeleteTagCommand deleteTagFromSelectedIndex = DeleteTagCommand.createWithSelectedIndex(getTypicalTags());
+
+        // same object -> returns true
+        assertEquals(deleteTagFromTargetIndexes, deleteTagFromTargetIndexes);
```
[the remark this answers is not in the document]
Comment (tanboonji) [reply src/test/java/seedu/address/logic/commands/DeleteTagCommandTest.java:167]: Added test cases for both different indexes and different tags.

**Review by oeiyiping — APPROVED** [review]
LGTM overall, but just a note:

When bulk deleting tags eg. `tag delete 1 2 3 -t test`, when only person 1 has the `test` tag, the output message will still show that 3 persons have had their `test` tag deleted.
Not sure if testers will consider this to be a bug (due to the inconsistency in actual effect vs output message), so maybe it would be good to change the output message to reflect a more accurate count, or include a note in the UG stating that this is intended output.

**Comment by tanboonji** [comment]
> LGTM overall, but just a note:
> 
> When bulk deleting tags eg. `tag delete 1 2 3 -t test`, when only person 1 has the `test` tag, the output message will still show that 3 persons have had their `test` tag deleted.
> Not sure if testers will consider this to be a bug (due to the inconsistency in actual effect vs output message), so maybe it would be good to change the output message to reflect a more accurate count, or include a note in the UG stating that this is intended output.

Will be looking at solving this bug in v1.4.

**Review by yaowei-soc — APPROVED** [review]
LGTM

**Pull request merged**